src/FastConv/FastConv.cpp

```diff
@@ -12,21 +12,190 @@ CFastConv::~CFastConv( void )
 
 Error_t CFastConv::init(float *pfImpulseResponse, int iLengthOfIr, int iBlockLength /*= 8192*/, ConvCompMode_t eCompMode /*= kFreqDomain*/)
 {
+    if (iLengthOfIr < 1) {
+        return Error_t::kFunctionInvalidArgsError;
+    }
 
+    m_ConvType = eCompMode;
+    m_pfImpulseResponse = new float [iLengthOfIr];
+
+    m_pfInputTail = new float[(2 * iLengthOfIr) - 1];
+    m_pfBlockBuffer = new float[iBlockLength];
+
+    m_iLengthOfIr = iLengthOfIr;
+    m_iblockSize = iBlockLength;
+
+    if (m_ConvType == CFastConv::ConvCompMode_t::kFreqDomain)
+    {
+        CFft::createInstance(m_pCFftInstance);
+        m_pCFftInstance->initInstance(2 * m_iblockSize, 1, CFft::kWindowHann, CFft::CFft::kPreWindow);
+    }
+
+    for (int i = 0; i < iLengthOfIr; i++)
+    {
+        m_pfImpulseResponse[i] = pfImpulseResponse[i];
+    }
+
+    m_bLongBlock = false;
     return Error_t::kNoError;
 }
 
 Error_t CFastConv::reset()
 {
+    delete[] m_pfImpulseResponse;
+    delete[] m_pfInputTail;
+    delete[] m_pfBlockBuffer;
+
+    m_bLongBlock = false;
+
+    if (m_ConvType == CFastConv::ConvCompMode_t::kFreqDomain)
+    {
+        m_pCFftInstance->resetInstance();
+        CFft::destroyInstance(m_pCFftInstance);
+    }
     return Error_t::kNoError;
 }
 
 Error_t CFastConv::process (float* pfOutputBuffer, const float *pfInputBuffer, int iLengthOfBuffers )
 {
-    return Error_t::kNoError;
+    if (m_iLengthOfIr < iLengthOfBuffers)
+    {
+        m_iTailIndex = iLengthOfBuffers - m_iLengthOfIr + 1;
+        for (int i = m_iTailIndex; i < iLengthOfBuffers + m_iLengthOfIr - 1; i++)
+        {
+            m_pfInputTail[i - m_iTailIndex] = (i < iLengthOfBuffers) ? pfInputBuffer[i] : 0;
+        }
+    }
+    else
+    {
+        m_iTailIndex = iLengthOfBuffers;
+        for (int i = 0; i < iLengthOfBuffers + m_iLengthOfIr - 1; i++)
+        {
+            m_pfInputTail[i] = (i < iLengthOfBuffers) ? pfInputBuffer[i] : 0;
+        }
+    }
+    
+    if (m_ConvType == CFastConv::ConvCompMode_t::kTimeDomain)
```

Comment:
This could be avoided by creating separate classes for the time and freq domain, and using CFastConv as a wrapper around them (This will allow you to save having an if condition being checked every time the process block is called)